Translation FAQ File Structure (faq-structure.md) (#98)
reactjs/ru.reactjs.org: content/docs/faq-structure.md

---
id: faq-structure
title: Структура файлов
permalink: docs/faq-structure.html
layout: docs
category: FAQ
---

> ### Как мне структурировать React-проекты? {#is-there-a-recommended-way-to-structure-react-projects}

Reviewer: ```suggestion
### Есть ли рекомендации по структуре React-проектов?
```

> Единого мнения на этот счёт не существует. Тем не менее, есть несколько общих подходов, популярных в экосистеме, которые вы, возможно, захотите рассмотреть.

Reviewer: ```suggestion
Единого мнения не существует. Однако, есть несколько популярных подходов, которые вы можете рассмотреть.
```

> #### Группировка по функционалу или маршруту {#grouping-by-features-or-routes}

Reviewer: ```suggestion
#### Группировка по возможностям или маршрутам {#grouping-by-features-or-routes}
```

> Один из популярных подходов -- это размещение файлов CSS, JS и тестов в папках, сгруппированных по возможности или маршруту.

Reviewer 2: @gcor, я знаю, что _feature_ в словаре определено как _возможность_, но в этом предложении это очень мешает.
В порядке исключения, предлагаю отойти от правила и вернуть неудобный _функционал_, который по контексту будет намного понятнее.

В текущем варианте (`...размещение файлов CSS, JS и тестов в папках, сгруппированных по возможности или маршруту`) очень сложно о каких возможностях речь. Первые два прочтения у меня вызывали ощущения, что это [наречное предложение](http://new.gramota.ru/spravka/punctum?layout=item&id=58_530).

@RinatRezyapov @LFedorov @lex111 какие ваши мнения?

Reviewer 3: > В порядке исключения, предлагаю отойти от правила и вернуть неудобный _функционал_, который по контексту будет намного понятнее.

:+1:

Reviewer: Тогда нужно «функциональность», потому что функционал это математический термин

Reviewer 3: Хорошее замечание!

Reviewer 2: @gcor и то правда! Чувствовал, что что-то не так, но не мог понять, что именно... Поправил в предложениях ниже.

> Один из популярных подходов -- это размещение файлов CSS, JS и тестов в папках, сгруппированных по возможности или маршруту.

Reviewer 2: ```suggestion
Один из популярных подходов -- это размещение файлов CSS, JS и тестов в папках, сгруппированных по функционалу или маршруту.
```

> Один из популярных подходов -- это размещение файлов CSS, JS и тестов в папках, сгруппированных по функционалу или маршруту.

Reviewer: ```suggestion
Один из популярных подходов -- это размещение файлов CSS, JS и тестов в папках, сгруппированных по возможности или маршруту.
```

  ProfileAPI.js
```

> Понятие «возможность» не является универсальным, поэтому выбор уровня детализации остается за вами. Если у вас не получается составить список папок верхнего уровня, вы можете спросить у пользователей вашего продукта, из каких основных частей он состоит, и взять модель мышления пользователей за образец.

Reviewer 2: Если решим сделать исключение, то надо не забыть исправить и тут _возможность_ на _функционал_.

> Понятие «возможность» не является универсальным, поэтому выбор уровня детализации остается за вами. Если у вас не получается составить список папок верхнего уровня, вы можете спросить у пользователей вашего продукта, из каких основных частей он состоит, и взять модель мышления пользователей за образец.

Reviewer 2: ```suggestion
Понятие «функционал» не является универсальным, поэтому выбор уровня детализации остается за вами. Если у вас не получается составить список папок верхнего уровня, вы можете спросить у пользователей вашего продукта, из каких основных частей он состоит, и взять модель мышления пользователей за образец.
```

Reviewer 3: Можно добавить как _синоним_, поскольку в некоторых случаях (как этот) предпочтительнее выбрать альтернативный вариант.

Author: _Возможность_ заменил на _функциональность_

> Понятие «функционал» не является универсальным и поэтому выбор уровня детализации остается за вами. Если у вас не получается составить список папок верхнего уровня, вы можете спросить у пользователей вашего продукта, из каких основных частей этот продукт состоит, и взять модель мышления пользователей за образец.

Reviewer: ```suggestion
Понятие «возможность» не является универсальным, поэтому выбор уровня детализации остается за вами. Если у вас не получается составить список папок верхнего уровня, вы можете спросить у пользователей вашего продукта, из каких основных частей он состоит, и взять модель мышления пользователей за образец.
```

[feature = возможность](https://github.com/reactjs/ru.reactjs.org/blob/master/TRANSLATION.md)

  ProfileHeader.js
  ProfileHeader.css
```

> Некоторые предпочитают идти еще дальше и размещать компоненты в различные папки в зависимости от их роли в приложении. К примеру, методология разработки [Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/) построена на этом принципе. Помните, что данные методологии следует рассматривать как полезные примеры, а не как строгие правила.

Reviewer 4: > Некоторые предпочитают идти еще дальше

ещё

#### Избегайте чрезмерной вложенности {#avoid-too-much-nesting}

> Проблем, связанных с чрезмерной вложенностью папок в JavaScript проектах, может возникнуть достаточно много. Одна из них -- это сложность контроля относительных импортов или обновления этих импортов при перемещении файлов. Если у вас нет веских оснований использовать глубокую вложенность папок, подумайте о том, чтобы ограничить себя максимум тремя или четырьмя уровнями вложенности в рамках одного проекта. Разумеется, это всего лишь рекомендация и она может быть не актуальна в конкретном случае.

Reviewer 3: ```suggestion
Проблем, связанных с чрезмерной вложенностью папок в JavaScript-проектах, может возникнуть достаточно много. Одна из них -- это сложность контроля относительных импортов или обновления этих импортов при перемещении файлов. Если у вас нет веских оснований использовать глубокую вложенность папок, подумайте о том, чтобы ограничить себя максимум тремя или четырьмя уровнями вложенности в рамках одного проекта. Разумеется, это всего лишь рекомендация и она может быть не актуальна в случае вашего проекта.
```

#### Не переусердствуйте {#dont-overthink-it}

> Если вы только начинаете проект, [не тратьте более 5 минут](https://ru.wikipedia.org/wiki/%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B9_%D0%BF%D0%B0%D1%80%D0%B0%D0%BB%D0%B8%D1%87) на выбор структуры проекта. Используйте любой из перечисленных выше подходов (или придумайте свой собственный) и начинайте писать код! Есть большая вероятность что вы вернетесь к переосмыслению структуры проекта после написания некоторого количества кода.

Reviewer 4: > Есть большая вероятность что вы вернетесь

Есть большая вероятность, что вы вернётесь

> Если вы только начинаете проект, [не тратьте более 5 минут](https://ru.wikipedia.org/wiki/%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B9_%D0%BF%D0%B0%D1%80%D0%B0%D0%BB%D0%B8%D1%87) на выбор структуры проекта. Используйте любой из перечисленных выше подходов (или придумайте свой собственный) и начинайте писать код! Есть большая вероятность, что вы вернётесь к переосмыслению структуры проекта после написания некоторого количества кода.

Reviewer 3: ```suggestion
Если вы только начинаете проект, [не тратьте более 5 минут](https://ru.wikipedia.org/wiki/%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B9_%D0%BF%D0%B0%D1%80%D0%B0%D0%BB%D0%B8%D1%87) на выбор структуры проекта. Выберите любой из перечисленных выше подходов (или придумайте свой собственный) и начните писать код! Есть большая вероятность, что вы вернётесь к переосмыслению структуры проекта после написания некоторого количества кода.
```

Author: Поправил

> Если вы чувствуете что окончательно застряли, начинайте с одной папки. Со временем, она станет настолько большой, что вам захочется отделить некоторые файлы от остальных. К этому времени у вас будет достаточно знаний, чтобы определить, какие файлы вы редактируете чаще всего. Как правило, файлы, которые часто меняются вместе, следует держать ближе друг к другу. Этот принцип называется «совместное размещение».

Reviewer: ```suggestion
Если вы чувствуете что окончательно застряли, начинайте с одной папки. Со временем, когда она станет настолько большой, что вам захочется отделить некоторые файлы от остальных, у вас будет достаточно знаний, чтобы определить, какие файлы вы редактируете чаще всего. Как правило, файлы, которые часто меняются вместе, следует держать ближе друг к другу. Этот принцип называется «совместное размещение».
```

Author: Поправил

> Если вы чувствуете что окончательно застряли, начинайте с одной папки. Со временем, она станет настолько большой, что вам захочется разделить некоторые файлы от остальных. К этому времени у вас будет достаточно знаний, чтобы определить, какие файлы вы редактируете чаще всего. Как правило, файлы, которые часто меняются вместе, следует держать ближе друг к другу. Этот принцип называется «совместное размещение».

Reviewer 4: > Со временем, она станет настолько большой, что вам захочется разделить некоторые файлы от остальных.

отделить?

Author: Поправил

> На практике, проекты часто используют сочетание нескольких вышеупомянутых подходов. Поэтому выбор «правильного» подхода, в самом начале проекта, не слишком важен.

Reviewer 3: ```suggestion
На практике проекты часто используют сочетание нескольких вышеупомянутых подходов. Поэтому выбор «правильного» подхода в самом начале проекта не слишком важен.
```

Author: Поправил
